go18_test.go

```diff
@@ -155,6 +156,32 @@ func TestContextCancelQuery(t *testing.T) {
 	}
 }
 
+// TestIssue617 tests that a failed query in QueryContext doesn't lead to a
+// goroutine leak.
+func TestIssue617(t *testing.T) {
+	db := openTestConn(t)
+	defer db.Close()
+
+	const N = 10
+
+	numGoroutineStart := runtime.NumGoroutine()
+	for i := 0; i < N; i++ {
+		query := `SELECT * FROM DOESNOTEXIST`
+		ctx, _ := context.WithTimeout(context.Background(), time.Second)
+		rows, err := db.QueryContext(ctx, query)
+		if rows != nil || err == nil {
```

Comment:
this is simultaneously testing too much and too little. There's no need to check that rows is nil - that's database/sql's API surface, not ours. However, we _should_ confirm that the error we get is the one we expect - it's not enough to check that the error is non-nil.